flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/generator/CirculantGraph.java

```diff
@@ -0,0 +1,161 @@
+	private long vertexCount;
+
+	private List<Tuple2<Long, Long>> startOffsetPairs = new ArrayList<>();
+
+	/**
+	 * An undirected {@link Graph} whose {@link Vertex} connects to targets appointed by an offset list.
+	 *
+	 * @param env the Flink execution environment
+	 * @param vertexCount number of vertices
+	 */
+	public CirculantGraph(ExecutionEnvironment env, long vertexCount) {
+		Preconditions.checkArgument(vertexCount >= MINIMUM_VERTEX_COUNT,
+			"Vertex count must be at least " + MINIMUM_VERTEX_COUNT);
+
+		this.env = env;
+		this.vertexCount = vertexCount;
+	}
+
+	/**
+	 * Required configuration for each offset of the graph.
+	 *
+	 * @param startOffset first offset appointing the vertices' position should be linked by any vertex
+	 * @param length offset in [startOffset, startOffset + length) will be added
+	 * @return this
+	 */
+	public CirculantGraph addOffsets(long startOffset, long length) {
+		long maxOffset = vertexCount / 2;
+		for (int i = 0; i < length; i++) {
+			long offset = startOffset + i;
+			Preconditions.checkArgument(offset >= MINIMUM_OFFSET,
+					"Offset must be at least " + MINIMUM_OFFSET);
+			Preconditions.checkArgument(offset <= maxOffset,
+					"Offset must be at most " + maxOffset);
+		}
+
+		// save startOffset and length pair
+		startOffsetPairs.add(new Tuple2<>(startOffset, length));
+
+		return this;
+	}
+
+	@Override
+	public Graph<LongValue, NullValue, NullValue> generate() {
+		// Vertices
+		DataSet<Vertex<LongValue, NullValue>> vertices = GraphGeneratorUtils.vertexSequence(env, parallelism, vertexCount);
+
+		// Edges
+		LongValueSequenceIterator iterator = new LongValueSequenceIterator(0, this.vertexCount - 1);
+
+		DataSet<Edge<LongValue, NullValue>> edges = env
+				.fromParallelCollection(iterator, LongValue.class)
+				.setParallelism(parallelism)
+				.name("Edge iterators")
+				.flatMap(new LinkVertexToOffsets(vertexCount, startOffsetPairs))
+				.setParallelism(parallelism)
+				.name("Circulant graph edges");
+
+		// Graph
+		return Graph.fromDataSet(vertices, edges, env);
+	}
+
+	@FunctionAnnotation.ForwardedFields("*->f0")
+	private static class LinkVertexToOffsets
+			implements FlatMapFunction<LongValue, Edge<LongValue, NullValue>> {
+
+		private final long vertexCount;
+
+		private final List<Tuple2<Long, Long>> startOffsetPairs;
+
+		private LongValue target = new LongValue();
+
+		private Edge<LongValue, NullValue> edge = new Edge<>(null, target, NullValue.getInstance());
+
+		public LinkVertexToOffsets(long vertexCount, List<Tuple2<Long, Long>> startOffsetPairs) {
+			this.vertexCount = vertexCount;
+			this.startOffsetPairs = startOffsetPairs;
+		}
+
+		@Override
+		public void flatMap(LongValue source, Collector<Edge<LongValue, NullValue>> out)
+				throws Exception {
+			edge.f0 = source;
+
+			// parse startOffsetPairs to offsets
+			List<Long> offsets = new ArrayList<>();
+			long maxOffset = vertexCount / 2;
+			for (Tuple2<Long, Long> offsetPair : startOffsetPairs) {
```

Comment:
There is no need to create an array of offsets, certainly not in every `flatMap`. Can simply iterate over the list of `Tuple2` and collect new edges.